src/main/go/org/jenkinsci/plugins/durabletask/heartbeat-launcher.go

```diff
@@ -0,0 +1,165 @@
+	"os"
+	"os/exec"
+	"os/signal"
+	"sync"
+
+	"golang.org/x/sys/unix"
+)
+
+// Note: EXIT signal is needed as some shell variants require an EXIT trap after catching a signal
+const trapSig = "trap ':' INT TERM EXIT"
+
+func checkHeartbeatErr(err error) {
+	checkErr("heartbeat", err)
+}
+
+func checkLauncherErr(err error) {
+	checkErr("launcher", err)
+}
+
+func checkErr(process string, err error) {
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "(%v) check err: %v\n", process, err.Error())
+	}
+}
+
+// Catch a termination signal to allow for a graceful exit (i.e. no zombies)
+func signalCatcher(sigChan chan os.Signal) {
+	for sig := range sigChan {
+		fmt.Printf("(sig catcher) caught: %v\n", sig)
+	}
+}
+
+func launcher(wg *sync.WaitGroup, pidChan chan int,
+	scriptString string, logPath string, resultPath string, outputPath string) {
+
+	defer wg.Done()
+	recordExit := fmt.Sprintf("status=\"$?\"; echo \"$status\" > %v.tmp; mv %v.tmp %v; wait; exit \"$status\"",
+		resultPath, resultPath, resultPath)
+	scriptWithExit := trapSig + "; " + scriptString + "; " + recordExit
+	scriptCmd := exec.Command("/bin/sh", "-c", scriptWithExit)
+	logFile, err := os.Create(logPath)
+	checkLauncherErr(err)
+	defer logFile.Close()
+
+	if outputPath != "" {
+		// capturing output
+		outputFile, err := os.Create(outputPath)
+		checkLauncherErr(err)
+		defer outputFile.Close()
+		scriptCmd.Stdout = outputFile
+	} else {
+		scriptCmd.Stdout = logFile
+	}
+	if outputPath != "" {
+		// capturing output
+		scriptCmd.Stderr = logFile
+	} else {
+		// Note: pointing to os.Stdout will not capture all err logs and fail unit tests
+		scriptCmd.Stderr = scriptCmd.Stdout
+	}
+	// Prevents script from being terminated if program gets terminated
+	scriptCmd.SysProcAttr = &unix.SysProcAttr{Setsid: true}
+	for i := 0; i < len(scriptCmd.Args); i++ {
+		fmt.Printf("(launcher) args %v: %v\n", i, scriptCmd.Args[i])
+	}
+	scriptCmd.Start()
+	pid := scriptCmd.Process.Pid
+	pidChan <- pid
+	fmt.Printf("(launcher) launched %v\n", pid)
+	err = scriptCmd.Wait()
+	checkLauncherErr(err)
+	resultVal := scriptCmd.ProcessState.ExitCode()
+	fmt.Printf("(launcher) script exit code: %v\n", resultVal)
+}
+
+func heartbeat(wg *sync.WaitGroup, launchedPid int,
+	controlDir string, resultPath string, logPath string) {
+
+	defer wg.Done()
+	const HBSCRIPT string = "heartbeat.sh"
+	fmt.Printf("(heartbeat) checking if %v is alive\n", launchedPid)
+	_, err := os.Stat(controlDir)
+	if os.IsNotExist(err) {
+		fmt.Fprintf(os.Stderr, "%v\n", err.Error())
+		return
+	}
+	_, err = os.Stat(resultPath)
+	if !os.IsNotExist(err) {
+		fmt.Printf("Result file already exists, stopping heartbeat.\n%v\n", resultPath)
+		return
+	}
+	// create the heartbeat script
+	heartbeat := fmt.Sprintf("#! /bin/sh\n%v; pid=\"$$\"; echo \"(heartbeat) pid: \"$pid\"\"; while true ; do kill -0 %v; status=\"$?\"; if [ \"$status\" -ne 0 ]; then break; fi; echo \"\"$pid\"\"; touch %v; sleep 3; done; echo \"(\\\"$pid\\\") exiting\"; exit",
```

Comment:
As above, this should be reimplemented as a goroutine or something. Part of the problem with the current implementation is that having the heartbeat done in a separate process is unreliable!

Reply:
I made a previous comment. tldr version: 1) Was always crashing as a goroutine 2)figured a workaround 3) discovered why it was crashing recently 4) didn't think about going back to goroutines 5) will now go back to giving goroutines a try